pkg/kubectl/history.go

```diff
@@ -0,0 +1,136 @@
+type HistoryViewer interface {
+	History(namespace, name string) (HistoryInfo, error)
+}
+
+func HistoryViewerFor(kind unversioned.GroupKind, c clientset.Interface) (HistoryViewer, error) {
+	switch kind {
+	case extensions.Kind("Deployment"):
+		return &DeploymentHistoryViewer{c}, nil
+	}
+	return nil, fmt.Errorf("no history viewer has been implemented for %q", kind)
+}
+
+// HistoryInfo stores the mapping from revision to podTemplate;
+// note that change-cause annotation should be copied to podTemplate
+type HistoryInfo struct {
+	RevisionToTemplate map[int64]*api.PodTemplateSpec
+}
+
+type DeploymentHistoryViewer struct {
+	c clientset.Interface
+}
+
+// History returns a revision-to-RC map as the revision history of a deployment
+func (h *DeploymentHistoryViewer) History(namespace, name string) (HistoryInfo, error) {
+	historyInfo := HistoryInfo{
+		RevisionToTemplate: make(map[int64]*api.PodTemplateSpec),
+	}
+	deployment, err := h.c.Extensions().Deployments(namespace).Get(name)
+	if err != nil {
+		return historyInfo, fmt.Errorf("failed to retrieve deployment %s: %v", name, err)
+	}
+	_, allOldRCs, err := deploymentutil.GetOldRCs(*deployment, h.c)
+	if err != nil {
+		return historyInfo, fmt.Errorf("failed to retrieve old RCs from deployment %s: %v", name, err)
+	}
+	newRC, err := deploymentutil.GetNewRC(*deployment, h.c)
+	if err != nil {
+		return historyInfo, fmt.Errorf("failed to retrieve new RC from deployment %s: %v", name, err)
+	}
+	allRCs := append(allOldRCs, newRC)
+	for _, rc := range allRCs {
+		v, err := deploymentutil.Revision(rc)
+		if err != nil {
+			return historyInfo, fmt.Errorf("failed to retrieve revision out of RC %s from deployment %s: %v", rc.Name, name, err)
+		}
+		historyInfo.RevisionToTemplate[v] = rc.Spec.Template
+		changeCause, err := getChangeCause(rc)
+		if err != nil {
+			return historyInfo, fmt.Errorf("failed to retrieve change-cause out of RC %s from deployment %s: %v", rc.Name, name, err)
+		}
+		if len(changeCause) > 0 {
+			if historyInfo.RevisionToTemplate[v].Annotations == nil {
+				historyInfo.RevisionToTemplate[v].Annotations = make(map[string]string)
+			}
+			historyInfo.RevisionToTemplate[v].Annotations[ChangeCauseAnnotation] = changeCause
+		}
+	}
+	return historyInfo, nil
+}
+
+// PrintRolloutHistory prints a formatted table of the input revision history of the deployment
+func PrintRolloutHistory(historyInfo HistoryInfo, resource, name string) (string, error) {
+	// Sort the revisionToChangeCause map by revision
+	var revisions []string
+	for k := range historyInfo.RevisionToTemplate {
+		revisions = append(revisions, strconv.FormatInt(k, 10))
+	}
+	sort.Strings(revisions)
+
+	return tabbedString(func(out io.Writer) error {
+		fmt.Fprintf(out, "%s %q:\n", resource, name)
+		fmt.Fprintf(out, "REVISION\tCHANGE-CAUSE\n")
+		errs := []error{}
+		for _, r := range revisions {
+			// Find the change-cause of revision r
+			r64, err := strconv.ParseInt(r, 10, 64)
```

Comment:
Well, I guess string->int64 is safer. But will we ever get to such numbers? Anyway, I am fine with this too